### ISSUE TITLE
compiler.py not handling None values

### ISSUE THREAD

**Issue by jt-helsinki** [sqlalchemy/sqlalchemy#4722]
I am using SQLAlchemy together with Pandas to insert dataframe data into a database (using pandas `dataframe.to_sql()`. Where values in the dataframe are `None`, SQLAlchemy still tries to run its data conversion functions in the `sql/compiler.py` class and subsequently throws errors.

For example, 
* the `quote_identifier()` function needs to return `'NULL'` if the value is `None`
* the `_escape_identifier()` function needs to return if the value is `None`. If not, the function throws an error complaining that replace does not exist on `NoneType`
* Same for `_unescape_identifier()`

I have manually hacked these changes locally and it solves the issue.

**Comment by zzzeek** [comment]
Hi there, and thanks for this issue !

Unfortunately I would require a lot more detail on this issue for it to be actionable, an overview can be seen in [Reporting Bugs](https://www.sqlalchemy.org/participate.html#bugs) on the SQLAlchemy website.   

I understand that it is tempting to dive into the source code and change things directly and to frame that as the only information necessary for the upstream maintainer (that's me) to act upon, but unfortunately things don't really work that way and the changes you have described mis-interpret the correct role of these functions, identifiers cannot be "None", and I cannot in my limited imagination conceive of a way that this could happen unless something was mis-using the API, which suggests perhaps a bug in Pandas to_sql() or in the way you are using it.   

Even if the changes you describe *were* appropriate, appreciate that SQLAlchemy has a test suite with well over 9000 tests that has been in development for 13 years, and it has never encountered a situation where None has been passed to these methods.    That is not to say it isn't happening, it is instead to mean, whatever you are doing is unusual enough that nobody has thought of it in 13 years, so being able to **see** how you've gotten this result is very important!  hence reproducing test code is essential, because I have no idea how None would get in there without doing something obviously wrong.

 I've responded inline below to try to illustrate further.

> I am using SQLAlchemy together with Pandas to insert dataframe data into a database (using pandas `dataframe.to_sql()`. 

How are these dataframes created?  Can you share a short Python script that reproduces what these dataframes look like?

> Where values in the dataframe are `None`, SQLAlchemy still tries to run its data conversion functions in the `sql/compiler.py` class and subsequently throws errors.

What is the error thrown?  Please post the complete stack trace in the issue.   Your program gave you this already so please share!


> 
> For example,
> 
>     * the `quote_identifier()` function needs to return `'NULL'` if the value is `None`
> 
>     * the `_escape_identifier()` function needs to return if the value is `None`. If not, the function throws an error complaining that replace does not exist on `NoneType`
> 
>     * Same for `_unescape_identifier()`

in SQL, an "identifier" is not a value inside of a column, it's a SQL keyword. If I gave you this statement:

   INSERT INTO my_table VALUES (?, ?)

Where are the "identifers"?   is it the "VALUES (? ?) "?   Actually no, it's the word "my_table", that's an identifier.   Suggesting the word NULL would be here means you are writing SQL like:

   INSERT INTO NULL VALUES (? ?)

That's not real SQL!   

Anyway, to sum up, somehow you are getting None into a place where pandas to_sql() expects the name of a column or a table.   This seems really broken and if you are using pandas correctly, then something is really broken in to_sql().  All in all, someone has to fix something, so please illustrate a script that shows how you are getting this result, thanks!

> 
> 
> I have manually hacked these changes locally and it solves the issue.

**Comment by jt-helsinki** [comment]
Thanks @zzzeek, I appreciate this detailed response. 

It seems this was my screw up. Originally I used:

```
engine = create_engine('mysql+pymysql://root:mypassword@localhost:6603', echo=True)
engine.execute("CREATE DATABASE IF NOT EXISTS ff") #create db
engine.execute("USE ff") # select new db
```

Instead, I see that I need to pass the database name in the connection string. 

```
engine = create_engine('mysql+pymysql://root:mypassword@localhost:6603', echo=True)
engine.execute("CREATE DATABASE IF NOT EXISTS ff") #create db
engine = create_engine('mysql+pymysql://root:mypassword@localhost:6603/ff', echo=True, pool_size=10, max_overflow=20)
```

Once I did this, it started working as expected.

**Comment by zzzeek** [comment]
heya -

OK that's great but I'd still be very curious to see how that possibly could have affected the identifier system.
